python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1019 @@
+# under the License.
+
+# cython: language_level = 3
+
+from __future__ import absolute_import
+
+from cython.operator cimport dereference as deref
+
+from pyarrow.lib cimport *
+from pyarrow.includes.libarrow_dataset cimport *
+from pyarrow.compat import frombytes, tobytes
+from pyarrow._fs cimport FileSystem, FileStats, Selector
+
+
+def _forbid_instantiation(klass, subclasses_instead=True):
+    msg = '{} is an abstract class thus cannot be initialized.'.format(
+        klass.__name__
+    )
+    if subclasses_instead:
+        subclasses = [cls.__name__ for cls in klass.__subclasses__]
+        msg += ' Use one of the subclasses instead: {}'.format(
+            ', '.join(subclasses)
+        )
+    raise TypeError(msg)
+
+
+cdef class FileFormat:
+
+    cdef:
+        shared_ptr[CFileFormat] wrapped
+        CFileFormat* format
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__)
+
+    cdef void init(self, const shared_ptr[CFileFormat]& sp):
+        self.wrapped = sp
+        self.format = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CFileFormat]& sp):
+        cdef FileFormat self
+
+        typ = frombytes(sp.get().name())
+        if typ == 'parquet':
+            self = ParquetFileFormat.__new__(ParquetFileFormat)
+        else:
+            raise TypeError(typ)
+
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CFileFormat] unwrap(self):
+        return self.wrapped
+
+    def name(self):
+        return frombytes(self.format.name())
+
+
+cdef class ParquetFileFormat(FileFormat):
+
+    def __init__(self):
+        self.init(shared_ptr[CFileFormat](new CParquetFileFormat()))
+
+
+cdef class PartitionScheme:
+
+    cdef:
+        shared_ptr[CPartitionScheme] wrapped
+        CPartitionScheme* scheme
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__)
+
+    cdef init(self, const shared_ptr[CPartitionScheme]& sp):
+        self.wrapped = sp
+        self.scheme = sp.get()
+
+    @staticmethod
+    cdef wrap(const shared_ptr[CPartitionScheme]& sp):
+        cdef PartitionScheme self
+
+        # TODO(kszucs): either choose type() or name() but consistently
+        typ = frombytes(sp.get().name())
+        if typ == 'schema_partition_scheme':
+            self = SchemaPartitionScheme.__new__(SchemaPartitionScheme)
+        elif typ == 'hive_partition_scheme':
+            self = HivePartitionScheme.__new__(HivePartitionScheme)
+        else:
+            raise TypeError(typ)
+
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CPartitionScheme] unwrap(self):
+        return self.wrapped
+
+    def parse(self, path):
+        cdef CResult[shared_ptr[CExpression]] result
+        result = self.scheme.Parse(tobytes(path))
+        return Expression.wrap(GetResultValue(result))
+
+
+cdef class SchemaPartitionScheme(PartitionScheme):
+
+    cdef:
+        CSchemaPartitionScheme* schema_scheme  # hmmm...
```

Comment:
Call this `concrete_scheme` if you want :-) (but then also change `HivePartitionScheme`)